File: src/material.rs
use crate::{
    surface::{HitResult, Ray},
    vec3::{Arrow, Color, color},
};

#[derive(Debug, Clone, Copy)]
pub struct MaterialResult {
    pub attenuation: Color,
    pub scattered: Ray,
}

pub trait Material {
    fn scatter(&self, ray: Ray, hit: HitResult) -> Option<MaterialResult>;
}

pub struct Lambert {
    pub albedo: Color,
}

impl Lambert {
    pub fn new(albedo: Color) -> Self {
        Self { albedo }
    }
}

impl Material for Lambert {
    fn scatter(&self, _ray: Ray, hit: HitResult) -> Option<MaterialResult> {
        let mut direction = hit.normal + Arrow::random_unit_vector();

        if direction.near_zero() {
            direction = hit.normal;
        }

        Some(MaterialResult {
            attenuation: self.albedo,
            scattered: Ray::new(hit.point, direction),
        })
    }
}

pub struct Metal {
    pub albedo: Color,
    pub fuzz: f64,
}

impl Metal {
    pub fn new(albedo: Color, fuzz: f64) -> Self {
        Self { albedo, fuzz }
    }
}

impl Material for Metal {
    fn scatter(&self, ray: Ray, hit: HitResult) -> Option<MaterialResult> {
        let reflected = ray.direction.reflect(hit.normal).unit_vector()
            + (self.fuzz * Arrow::random_unit_vector());
        let scattered = Ray::new(hit.point, reflected);

        (scattered.direction.dot(hit.normal) > 0.0).then_some(MaterialResult {
            attenuation: self.albedo,
            scattered,
        })
    }
}

pub struct Dielectric {
    pub refraction_index: f64,
}

impl Dielectric {
    pub fn new(refraction_index: f64) -> Self {
        Self { refraction_index }
    }
}

impl Material for Dielectric {
    fn scatter(&self, ray: Ray, hit: HitResult) -> Option<MaterialResult> {
        let refraction_index = if hit.front_face {
            1.0 / self.refraction_index
        } else {
            self.refraction_index
        };

        let unit_direction = ray.direction.unit_vector();
        let cos_theta = -unit_direction.dot(hit.normal).min(1.0);
        let sin_theta = (1.0 - cos_theta * cos_theta).sqrt();

        let direction = if refraction_index * sin_theta > 1.0 {
            unit_direction.reflect(hit.normal)
        } else {
            unit_direction.refract(hit.normal, refraction_index)
        };

        Some(MaterialResult {
            attenuation: color(1.0, 1.0, 1.0),
            scattered: Ray::new(hit.point, direction),
        })
    }
}

File: raytracer_ts/material.ts
import { Ray, type HitResult } from './hittable';
import { type Texture } from './texture';
import { Interval, Vec3, type Color3, type Point3 } from './util';

export class MaterialResult {
    constructor(
        public attenuation: Color3,
        public scattered: Ray,
    ) {}
}

export abstract class Material {
    public abstract scatter(ray: Ray, hit: HitResult): MaterialResult | null;

    public emitted(u: number, v: number, point: Point3) {
        return Vec3.zero;
    }
}

export class Lambert extends Material {
    constructor(private texture: Texture) {
        super();
    }

    public override scatter(ray: Ray, hit: HitResult): MaterialResult | null {
        let scatterDirection = hit.normal.plus(Vec3.randomUnitVector());

        if (scatterDirection.nearZero) {
            scatterDirection = hit.normal;
        }

        const scattered = new Ray(hit.point, scatterDirection, ray.time);
        const attenuation = this.texture.value(hit.u, hit.v, hit.point);
        return new MaterialResult(attenuation, scattered);
    }
}

export class Metal extends Material {
    constructor(
        private albedo: Color3,
        private fuzz: number,
    ) {
        super();
    }

    public override scatter(ray: Ray, hit: HitResult): MaterialResult | null {
        const reflected = ray.direction.reflect(hit.normal);
        const fuzzed = reflected.unitVector.plus(Vec3.randomUnitVector().times(this.fuzz));
        const scattered = new Ray(hit.point, fuzzed, ray.time);

        if (scattered.direction.dot(hit.normal) > 0) {
            return new MaterialResult(this.albedo, scattered);
        }

        return null;
    }
}

export class Dielectric extends Material {
    constructor(private refractionIndex: number) {
        super();
    }

    public scatter(ray: Ray, hit: HitResult): MaterialResult | null {
        const refractionIndex = hit.frontFace ? 1.0 / this.refractionIndex : this.refractionIndex;

        const unitDirection = ray.direction.unitVector;
        const cosTheta = Math.min(unitDirection.neg.dot(hit.normal), 1.0);
        const sinTheta = Math.sqrt(1.0 - cosTheta ** 2);

        const cannotRefract = refractionIndex * sinTheta > 1.0;

        const direction =
            cannotRefract || this.reflectance(cosTheta) > Interval.unit.random()
                ? unitDirection.reflect(hit.normal)
                : unitDirection.refract(hit.normal, refractionIndex);

        return new MaterialResult(Vec3.one, new Ray(hit.point, direction, ray.time));
    }

    private reflectance(cosine: number) {
        const r0 = ((1 - this.refractionIndex) / (1 + this.refractionIndex)) ** 2;
        return r0 + (1 - r0) * (1 - cosine) ** 5;
    }
}

export class DiffuseLight extends Material {
    constructor(private texture: Texture) {
        super();
    }

    public scatter(ray: Ray, hit: HitResult): MaterialResult | null {
        return null;
    }

    public emitted(u: number, v: number, point: Point3): Vec3 {
        return this.texture.value(u, v, point);
    }
}
